feat(api,robot-server): Overhaul robot calibration flows

sfoster1 opened the pull request:
This commit makes the calibration overhaul the new way to calibrate the
OT-2.

In the new system, there are three components to robot calibration:

1. Deck attitude calibration is a 3 by 3 linear transform that maps the
attitude of the deck relative to the gantry. By convention, the
Z-coupled parts of the transform are identity, and only the in-plane
parts are kept. Deck attitude is calibrated using the deck calibration
session, which can be initiated with a POST to /sessions with session
type deckCalibration. In this session, the pipette is jogged to the deck
in slot 5 to establish a baseline Z position for the rest (the value is
not saved), then to the calibration crosses in slots 1, 3, and 7. The XY
position of these crosses is saved and regressed onto the default
positions for the crosses to form the deck calibration. There is no
longer any offset encoded into deck calibration, and it is applied in
the same way to both mounts.

2. Tip length calibration is a single number, representing the distance
between the end of a particular kind of tip (by labware definition hash
or labware URI) and the end of a particular pipette's (by pipette serial
number) nozzle when attached to that pipette. This replaces part of what
the tip probe flow does and is entirely manual, allowing fine tuning and
removing the possibility of certain kinds of catastrophic error in
calibration. You can start a tip length calibration session with a POST
to /sessions with session type tipLengthCalibration. This kind of
calibration is most often done in the pre-protocol flow, when it is
executed for every tiprack and pipette combination. It is calibrated by
jogging the end of a pipette nozzle to a fixed point, then jogging the
end of a tip attached to that nozzle to a fixed point, and saving the
difference in position in Z.

3. Pipette offset calibration is an XYZ offset, representing the vector
between the nominal position of the end of a nozzle of a particular
pipette (by pipette serial number) on a particular mount (by mount id)
and its actual position. It is calibrated by jogging the pipette, with a
calibrated tip attached, to the deck in slot 5 to capture the Z, then to
the calibration cross in slot 1 to capture the X and Y. Because a tip is
required, clearing or re-calibrating the kind of tip used for a specific
pipette offset calibration requires re-calibrating that pipette. A
pipette offset calibration session can be started with a POST to
/sessions with session type pipetteOffsetCalibration. If there is no
calibrated tip suitable for pipette offset calibration on the specified
pipette, the session will go through tip length calibration for that
tip before moving on to pipette offset calibration. This behavior can
also be forced with a session create parameter.

In addition, pipette sizes are now described in absolute terms from the
mount screws to the end of the nozzle, rather than being relative to a
P300 Single GEN1. This data is captured in the new nozzle_offset value
in the pipette configuration.

The legacy calibration (/calibrate/deck and friends, the probe RPC
endpoint, the CLI version of deck calibration) have been removed and the
internals switched to use the new kind of calibration only.

Closes #6738, #6739


## Review Requests
- Make sure I didn't break anything

sfoster1 committed:
feat(api,robot-server): Overhaul robot calibration flows

This commit makes the calibration overhaul the new way to calibrate the
OT-2.

In the new system, there are three components to robot calibration:

1. Deck attitude calibration is a 3 by 3 linear transform that maps the
attitude of the deck relative to the gantry. By convention, the
Z-coupled parts of the transform are identity, and only the in-plane
parts are kept. Deck attitude is calibrated using the deck calibration
session, which can be initiated with a POST to /sessions with session
type deckCalibration. In this session, the pipette is jogged to the deck
in slot 5 to establish a baseline Z position for the rest (the value is
not saved), then to the calibration crosses in slots 1, 3, and 7. The XY
position of these crosses is saved and regressed onto the default
positions for the crosses to form the deck calibration. There is no
longer any offset encoded into deck calibration, and it is applied in
the same way to both mounts.

2. Tip length calibration is a single number, representing the distance
between the end of a particular kind of tip (by labware definition hash
or labware URI) and the end of a particular pipette's (by pipette serial
number) nozzle when attached to that pipette. This replaces part of what
the tip probe flow does and is entirely manual, allowing fine tuning and
removing the possibility of certain kinds of catastrophic error in
calibration. You can start a tip length calibration session with a POST
to /sessions with session type tipLengthCalibration. This kind of
calibration is most often done in the pre-protocol flow, when it is
executed for every tiprack and pipette combination. It is calibrated by
jogging the end of a pipette nozzle to a fixed point, then jogging the
end of a tip attached to that nozzle to a fixed point, and saving the
difference in position in Z.

3. Pipette offset calibration is an XYZ offset, representing the vector
between the nominal position of the end of a nozzle of a particular
pipette (by pipette serial number) on a particular mount (by mount id)
and its actual position. It is calibrated by jogging the pipette, with a
calibrated tip attached, to the deck in slot 5 to capture the Z, then to
the calibration cross in slot 1 to capture the X and Y. Because a tip is
required, clearing or re-calibrating the kind of tip used for a specific
pipette offset calibration requires re-calibrating that pipette. A
pipette offset calibration session can be started with a POST to
/sessions with session type pipetteOffsetCalibration. If there is no
calibrated tip suitable for pipette offset calibration on the specified
pipette, the session will go through tip length calibration for that
tip before moving on to pipette offset calibration. This behavior can
also be forced with a session create parameter.

In addition, pipette sizes are now described in absolute terms from the
mount screws to the end of the nozzle, rather than being relative to a
P300 Single GEN1. This data is captured in the new nozzle_offset value
in the pipette configuration.

The legacy calibration (/calibrate/deck and friends, the probe RPC
endpoint, the CLI version of deck calibration) have been removed and the
internals switched to use the new kind of calibration only.

Closes #6738, #6739

Laura-Danielle reviewed (approved):
beautiful 🔥  🚒  🧑‍🚒  🧨 

tested on my robot, all the flows worked as expected.

Laura-Danielle commented:
Can we also remove the `tip_probe` function in `api/calibration.py`?

sfoster1 committed:
fixup: fix some user flow logic around cal block

The cal block logic should be fused between the flag and whether or not
the calblock is loaded; and also, it should be loaded based on the
eventually-decided-on recalibration flag, not just the create param.
